File: moxy-compiler/src/main/java/com/omegar/mvp/compiler/pipeline/Processor.java
package com.omegar.mvp.compiler.pipeline;

/**
 * Created by Anton Knyazev on 03.12.2020.
 */
public abstract class Processor<I,O> {

    public void process(I input, PipelineContext<O> context) {
        context.next(process(input));
    }

    protected O process(I input) {
        //noinspection unchecked
        return (O) input;
    }

    protected void finish() {
        // nothing
    }

}

File: moxy/compiler/src/main/java/com/omegar/mvp/compiler/pipeline/Processor.kt
package com.omegar.mvp.compiler.pipeline

/**
 * Created by Anton Knyazev on 03.12.2020.
 */
abstract class Processor<I, O> {

    open fun process(input: I, context: PipelineContext<O>?) {
        context?.next(process(input))
    }

    internal open fun process(input: I): O {
        return input as O
    }

    open fun finish(nextContext: PipelineContext<O>?) {
        nextContext?.finish()
    }

    fun withCache(): Processor<I, O> {
        return CacheWrapperProcessor(this)
    }
}